File: lib/openssl/keylog/session_extensions.rb
require 'openssl/keylog'

module OpenSSL::Keylog
  module SessionExtensions
    def to_keylog
      # TODO: Create an actual implementation.
      'CLIENT_RANDOM RANDOM HOSTKEY'
    end
  end
end

OpenSSL::SSL::Session.send(:include, OpenSSL::Keylog::SessionExtensions)

File: ext/session_extensions/session_extensions.c
#include <ruby.h>

static VALUE
to_keylog(VALUE self)
{
  return rb_str_new2("CLIENT_RANDOM RANDOM HOSTKEY");
}

void
Init_session_extensions()
{
  VALUE mOpenSSL = rb_const_get(rb_cObject, rb_intern("OpenSSL"));
  VALUE mKeylog  = rb_const_get(mOpenSSL, rb_intern("Keylog"));

  VALUE mSessionExtensions = rb_define_module_under(mKeylog, "SessionExtensions");

  rb_define_method(mSessionExtensions, "to_keylog", to_keylog, 0);
}
